src/audio/host_zephyr.c

```diff
@@ -0,0 +1,1135 @@
+	dev->state = COMP_STATE_READY;
+
+	return dev;
+}
+
+static void host_free(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+
+	comp_info(dev, "host_free()");
+
+	dma_put(hd->dma);
+
+	ipc_msg_free(hd->msg);
+	dma_sg_free(&hd->config.elem_array);
+	rfree(hd);
+	rfree(dev);
+}
+
+static int host_elements_reset(struct comp_dev *dev)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_elem *source_elem = NULL;
+	struct dma_sg_elem *sink_elem = NULL;
+	struct dma_sg_elem *local_elem;
+
+	/* setup elem to point to first source elem */
+	source_elem = hd->source->elem_array.elems;
+	if (source_elem) {
+		hd->source->current = 0;
+		hd->source->current_end = source_elem->src + source_elem->size;
+	}
+
+	/* setup elem to point to first sink elem */
+	sink_elem = hd->sink->elem_array.elems;
+	if (sink_elem) {
+		hd->sink->current = 0;
+		hd->sink->current_end = sink_elem->dest + sink_elem->size;
+	}
+
+	/* local element */
+	if (source_elem && sink_elem) {
+		local_elem = hd->config.elem_array.elems;
+		local_elem->dest = sink_elem->dest;
+		local_elem->size =
+			dev->direction == SOF_IPC_STREAM_PLAYBACK ?
+			sink_elem->size : source_elem->size;
+		local_elem->src = source_elem->src;
+	}
+
+	return 0;
+}
+
+static int host_verify_params(struct comp_dev *dev,
+			      struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_dbg(dev, "host_verify_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "host_verify_params(): comp_verify_params() failed");
+		return ret;
+	}
+
+	return 0;
+}
+
+/* configure the DMA params and descriptors for host buffer IO */
+static int host_params(struct comp_dev *dev,
+		       struct sof_ipc_stream_params *params)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_config *config = &hd->config;
+	struct dma_sg_elem *sg_elem;
+	struct dma_config *dma_cfg = &hd->z_config;
+	struct dma_block_config dma_block_cfg;
+	uint32_t period_count;
+	uint32_t period_bytes;
+	uint32_t buffer_size;
+	uint32_t addr_align;
+	uint32_t align;
+	int i, channel, err;
+
+	comp_dbg(dev, "host_params()");
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+		hd->local_buffer = list_first_item(&dev->bsink_list,
+						   struct comp_buffer,
+						   source_list);
+	else
+		hd->local_buffer = list_first_item(&dev->bsource_list,
+						   struct comp_buffer,
+						   sink_list);
+
+	err = host_verify_params(dev, params);
+	if (err < 0) {
+		comp_err(dev, "host_params(): pcm params verification failed.");
+		return -EINVAL;
+	}
+
+	/* host params always installed by pipeline IPC */
+	hd->host_size = params->buffer.size;
+	hd->stream_tag = params->stream_tag;
+	hd->no_stream_position = params->no_stream_position;
+	hd->host_period_bytes = params->host_period_bytes;
+	hd->cont_update_posn = params->cont_update_posn;
+
+	/* retrieve DMA buffer address alignment */
+	err = dma_get_attribute(hd->dma, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
```

Comment:
dma_get_attrubite is still missing in new version, Zephyr API does not support this.
Not sure if we can use here defines directly?